protected/tests/unit/models/AUSPatientTest.php

```diff
@@ -18,45 +18,244 @@ public function tearDown()
+
+        return $new_pat;
+    }
+
+    //create a GP with everything in order
+    protected function getGP()
+    {
+        $new_gp = new Gp();
+        $new_gp->obj_prof = $this->getNDigitRand(20);
+        $new_gp->nat_id = $this->getNDigitRand(20);
+        $new_gp->save();
+        return $new_gp;
+    }
+
+    protected function getNDigitRand($noDigits = 40)
     {
-        //create a patient as a duplicate of one in the database
-        $patient_dup = new Patient;
-        $patient_dup->attributes = $this->patient['patient1'];
+        $rand_string = mt_rand(1,9);
+        for ($i = 1; $i < $noDigits; $i++)
+        {
+            $rand_string .= mt_rand(0,9);
+        }
+        return $rand_string;
+    }
 
-        //change the things we don't want it to fail on
-        $patient_dup->id = NULL;
-        $patient_dup->pas_key = NULL;
-        $patient_dup->hos_num = 65465465;
+    public function testUniqueMedicareNo()
+    {
+        $patient_dup = $this->getUnsavedNewPatient('patient1');
+        $patient_dup->nhs_num = $this->patient['patient1']['nhs_num'];
 
         //fail to save with duplication
         $this->assertFalse($patient_dup->save());
         $this->assertSame('Medicare Number "" has already been taken.' //remove the actual number
             , preg_replace('/\d+/', '' , $patient_dup->getErrors('nhs_num')[0]));
 
         //remove duplication and save successfully
-        $patient_dup->nhs_num = 65465445;
+        $patient_dup->nhs_num = $this->getNDigitRand();
         $this->assertTrue($patient_dup->save());
     }
 
     public function testUniqueHosNo()
     {
-        //create a patient as a duplicate of one in the database
-        $patient_dup = new Patient;
-        $patient_dup->attributes = $this->patient['patient1'];
-
-        //change the things we don't want it to fail on
-        $patient_dup->id = NULL;
-        $patient_dup->pas_key = NULL;
-        $patient_dup->nhs_num = 65465465;
+        $patient_dup = $this->getUnsavedNewPatient('patient1');
+        $patient_dup->hos_num = $this->patient['patient1']['hos_num'];
 
         //fail to save with duplication
         $this->assertFalse($patient_dup->save());
         $this->assertSame('CERA Number "" has already been taken.'
             , preg_replace('/\d+/', '' , $patient_dup->getErrors('hos_num')[0]));
 
         //remove duplication and save successfully
-        $patient_dup->hos_num = 65465445;
+        $patient_dup->hos_num = $this->getNDigitRand();
         $this->assertTrue($patient_dup->save());
     }
+
+    public function testHosNoRequired()
+    {
+        $new_pat = $this->getUnsavedNewPatient('patient1');
+
+        //Remove the CERA/Hospital number
+        $new_pat->hos_num = NULL;
+
+        //Check that this fails and in the right manner
+        $this->assertFalse($new_pat->save());
+        $this->assertSame('CERA Number cannot be blank.', $new_pat->getError('hos_num'));
```

Comment:
We should all start using assertSame more often it seems (https://stackoverflow.com/questions/10254180/)

Reply:
Did not even know that was a thing 😲